accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,474 @@
+
+Today, a managed pointer is obtained safely through the `ref` keyword in C#. It references a storage location of an object or value type which can be a stack variable, static variable, parameter, field or array element. If the storage location is a field or array element, the managed pointer is referred to as an "interior pointer" which is supported by GC meaning that GC won't collect the owning object even if there are only interior pointers to it.
+
+A managed pointer can becreated in several ways:
+- [System.TypedReference](https://learn.microsoft.com/en-us/dotnet/api/system.typedreference). Since this a byref-like type, it is only stack-allocated. Internally, it uses a `ref byte` approach along with a reference to a `Type`. Note `TypeReference` is a special type and has its own opcodes (mkrefany, refanytype, refanyval) which translate to C# keyworkds (`__makeref`, `__reftype`, `__refvalue`).
+- Using `ref <T>` for the strongly typed case or `ref byte` for the loosely-typed case. This is expanded in 7.0 due to the new "ref field" support. Previously, there was an internal `ByReference<T>` class that was used and in 7.0 this was changed to `ByReference` in 8.0 which no longer maintains the `<T>` type and internally just contains `ref byte`. This `ByReference` type is used today in reflection invoke when there are <=4 parameters.
+- Using unsafe `void*` (or `IntPtr`) with GC tracking or pinning to make the use GC-safe. Tracking is supported internally through the use of a newer "RegisterForGCReporting()" mechanism. This approach is used today in reflection invoke when there are >=5 parameters.
+
+# Proposed APIs
+
+## MethodInvoker
+This ref struct is the mechanism to specify the target + arguments (including return value) and supports these mechanisms:
+- `object` (including boxing). Supports loose coupling scenarios are supported, like reflection today.
+- `ref <T>`. Supports new scenarios as mentioned earlier; type must be known ahead-of-time and due to no language support, cannot be a byref-like type like `Span<T>`.
+- `void*`. Unsafe cases used to support byref-like types in an unsafe manner.
+- `TypedReference`. Optional for now; pending language asks, it may make supporting byref-like types a safe operation.
+
+```cs
+namespace System.Reflection
+{
+     public ref struct MethodInvoker
+     {
+        // Zero-arg case:
+        public MethodInvoker()
+
+        // Variable-length number of arguments:
+        public unsafe MethodInvoker(ArgumentValue* argumentStorage, int argCount)
+
+        // Fixed length (say up to 8)
+        public MethodInvoker(ref ArgumentValuesFixed values)
+
+        // Dispose needs to be called with variable-length case
+        public void Dispose()
```

Comment:
I'm wondering if having Dispose which is only required in some cases won't trigger false positives in disposable analyzers.

Reply:
Good question.